haystack/nodes/label_generator/pseudo_label_generator.py

```diff
@@ -0,0 +1,216 @@
+        :type question_producer: Union[QuestionGenerator, List[Dict[str, str]]]
+        :param retriever: The retriever used to query document stores
+        :type retriever: BaseRetriever
+        :param cross_encoder_model_name_or_path: The path to the cross encoder model, defaults to
+        cross-encoder/ms-marco-MiniLM-L-6-v2
+        :type cross_encoder_model_name_or_path: str (optional)
+        :param total_number_of_questions: This is the total number of questions that will be generated, defaults
+        to sys.maxsize
+        :type total_number_of_questions: int
+        :param top_k: This is the number of answers that you want to retrieve for each question, defaults to 10
+        :type top_k: int (optional)
+        """
+
+        super().__init__()
+        self.question_document_pairs = None
+        self.question_generator = None
+        if isinstance(question_producer, QuestionGenerator):
+            self.question_generator = question_producer
+        elif isinstance(question_producer, list) and len(question_producer) > 0:
+            example = question_producer[0]
+            if isinstance(example, dict) and "question" in example and "document" in example:
+                self.question_document_pairs = question_producer
+            else:
+                raise ValueError("question_producer list must contain dicts with keys 'question' and 'document'")
+        else:
+            raise ValueError("Provide either a QuestionGenerator or nonempty list of questions/document pairs")
+
+        self.retriever = retriever
+        self.cross_encoder = CrossEncoder(cross_encoder_model_name_or_path)
+        self.total_number_of_questions = total_number_of_questions
+        self.top_k = top_k
+
+    def generate_questions(self, documents: List[Document]) -> List[Dict[str, str]]:
+        """
+        It takes a list of documents and generates a list of question-document pairs
+
+        :param documents: A list of documents to generate questions from
+        :type documents: List[Document]
+        :return: A list of question-document pairs.
+        """
+        question_doc_pairs: List[Dict[str, str]] = []
+        if self.question_document_pairs:
+            question_doc_pairs = self.question_document_pairs
+        else:
+            done = False
+            for idx, doc in enumerate(documents):
+                questions = self.question_generator.generate(text=doc.content)
+                for q in questions:
+                    question_doc_pairs.append({"question": q.strip(), "document": doc.content})
+                    done = len(question_doc_pairs) >= self.total_number_of_questions
+                    if done:
+                        break
+                if done:
+                    break
+        return question_doc_pairs
+
+    def mine_negatives(self, question_doc_pairs: List[Dict[str, str]], batch_size: int = 1) -> List[Dict[str, str]]:
```

Comment:
I am not sure if maintaining a fixed set of question-positive-negative triplets has an effect on the outcome of the training. I think that for GPL they randomly sample dynamically question-positive-negative triplets and calculate the margin on the fly (the CrossEncoder scores are still precomputed). You can see an example of this implementation in Nils Reimers' MSMarcoDataset (https://github.com/UKPLab/sentence-transformers/blob/40af04ed70e16408f466faaa5243bee6f476b96e/examples/training/ms_marco/train_bi-encoder_margin-mse.py#L178). This results in the model seeing different triplets for each epoch which I suspect has quite some impact on the training results.

Reply:
I followed the GPL [notebook](https://colab.research.google.com/gist/jamescalam/d2c888775c87f9882bb7c379a96adbc8/gpl-domain-adaptation.ipynb#scrollTo=ad4b870e) Niels demoed for Pinecone.

Reply:
Ok, valid point. Maybe also check how it's done in the official GPL repo https://github.com/UKPLab/gpl . I'm sure that the demo notebook is a bit simplified when compared to the paper.